## Global Analysis

### Import Libraries

In [5]:
import pandas as pd

### Load Datasets

The John Hopkins University provides an updated repository of the Coronavirus Pandemic Datasets.
<br>
The dataset is maintained and the update frequency is daily at about 23:59 UTC. 

### Task 1
<br>
The task here is to read the data from the John Hopkins University GitHub repository, and save it to your workspace for analysis.
<br>
Repository - [Link](https://raw.githubusercontent.com/datasets/covid-19/master/data/time-series-19-covid-combined.csv)
<br>
You can check out this guide [here](https://community.insaid.co/hc/en-us/articles/360052303453-How-to-read-a-csv-dataset-from-GitHub-in-Pandas-) on how to read the data


In [7]:
global_cases = 'https://raw.githubusercontent.com/datasets/covid-19/master/data/time-series-19-covid-combined.csv'

In [8]:
df_global_cases = pd.read_csv(global_cases)

### Task 2
<br>
Using the Pandas `head` method, view the first five rows

In [9]:
df_global_cases.head()

Date Country/Region Province/State  Confirmed  Recovered  Deaths
0  2020-01-22    Afghanistan            NaN          0        0.0       0
1  2020-01-23    Afghanistan            NaN          0        0.0       0
2  2020-01-24    Afghanistan            NaN          0        0.0       0
3  2020-01-25    Afghanistan            NaN          0        0.0       0
4  2020-01-26    Afghanistan            NaN          0        0.0       0

## Data Preparation

### Task 3
<br>
* Rename the `Country/Region` column to `country`
<br>
* Drop the `Province/State` column 

In [12]:
df_global_cases = df_global_cases.rename(columns={'Country/Region':'Country'})

In [14]:
df_global_cases = df_global_cases.drop('Province/State', axis=1)

In [15]:
df_global_cases.head()

Date      Country  Confirmed  Recovered  Deaths
0  2020-01-22  Afghanistan          0        0.0       0
1  2020-01-23  Afghanistan          0        0.0       0
2  2020-01-24  Afghanistan          0        0.0       0
3  2020-01-25  Afghanistan          0        0.0       0
4  2020-01-26  Afghanistan          0        0.0       0

### Task 4
<br>
* It is of interest to determine the number of active cases of the virus. Your task here is to calculate the active cases by finding the differences between current confirmed cases and non-confirmed cases(Recovered & Death)
<br>
* Save your result to a new column - `Active`

In [16]:
df_global_cases['Active'] = df_global_cases['Confirmed'] - df_global_cases['Recovered'] - df_global_cases['Deaths']

In [22]:
df_global_cases.tail()

Date   Country  Confirmed  Recovered  Deaths  Active
87528  2020-12-05  Zimbabwe      10617     8844.0     291  1482.0
87529  2020-12-06  Zimbabwe      10718     8880.0     291  1547.0
87530  2020-12-07  Zimbabwe      10839     8972.0     294  1573.0
87531  2020-12-08  Zimbabwe      10912     9062.0     303  1547.0
87532  2020-12-09  Zimbabwe      11007     9147.0     304  1556.0

### Task 5 
* Explore the dataset by picking a country e.g `Afghanistan`, and determine the total number of cases in the country.
<br>
<br> 
* You will observe some anomalies, write your observation, and how you intend to fix this. 

In [25]:
df_global_cases[df_global_cases['Country']=='Afghanistan'].sum()

Date             2020-01-222020-01-232020-01-242020-01-252020-0...
Country          AfghanistanAfghanistanAfghanistanAfghanistanAf...
Confirmed                                                  7387702
Recovered                                              5.12811e+06
Deaths                                                      251784
Active                                                 2.00781e+06
confirmed_new                                                47716
dtype: object

### Observation 
The total number cases is a lot e.g The confirmed cases is 7387702, which is way more than the total confirmed cases on 2020-12-09. A quick google search shows the total value to be about 47500.

### Task 6
<br>
* From your observation, create new columns for `Confirmed`, `Recovered`, `Deaths` and `Active`
<br>
<br>
Hint
<br>
<br>
* Group by `Country`, and specify the columns on which you want to perform the actual aggregation.
<br>
<br>
* Shift each group by periods observations.
<br>
<br>
* Subtract the cumulative values

In [26]:
df_global_cases['confirmed_new'] = df_global_cases.groupby('Country')['Confirmed'].shift().diff(1)
df_global_cases['recovered_new'] = df_global_cases.groupby('Country')['Recovered'].shift().diff(1)
df_global_cases['deaths_new'] = df_global_cases.groupby('Country')['Deaths'].shift().diff(1)
df_global_cases['active_new'] = df_global_cases.groupby('Country')['Active'].shift().diff(1)

In [27]:
df_global_cases.head()

Date      Country  Confirmed  Recovered  Deaths  Active  \
0  2020-01-22  Afghanistan          0        0.0       0     0.0   
1  2020-01-23  Afghanistan          0        0.0       0     0.0   
2  2020-01-24  Afghanistan          0        0.0       0     0.0   
3  2020-01-25  Afghanistan          0        0.0       0     0.0   
4  2020-01-26  Afghanistan          0        0.0       0     0.0   

   confirmed_new  recovered_new  deaths_new  active_new  
0            NaN            NaN         NaN         NaN  
1            NaN            NaN         NaN         NaN  
2            0.0            0.0         0.0         0.0  
3            0.0            0.0         0.0         0.0  
4            0.0            0.0         0.0         0.0

### Task 7 
* Drop the old columns

In [28]:
df_global_cases = df_global_cases.drop(['Confirmed', 'Recovered', 'Deaths', 'Active'], axis = 1)

In [29]:
df_global_cases.head()

Date      Country  confirmed_new  recovered_new  deaths_new  \
0  2020-01-22  Afghanistan            NaN            NaN         NaN   
1  2020-01-23  Afghanistan            NaN            NaN         NaN   
2  2020-01-24  Afghanistan            0.0            0.0         0.0   
3  2020-01-25  Afghanistan            0.0            0.0         0.0   
4  2020-01-26  Afghanistan            0.0            0.0         0.0   

   active_new  
0         NaN  
1         NaN  
2         0.0  
3         0.0  
4         0.0

### Task 8
* Create a new table to calculate the total number of cases by country
* Save your result to a new table - `total_cases`

In [31]:
total_cases = df_global_cases.groupby('Country').sum()

In [32]:
total_cases

confirmed_new  recovered_new  deaths_new  active_new
Country                                                                 
Afghanistan               47716.0        37920.0      1906.0      7890.0
Albania                   44436.0        22527.0       936.0     20973.0
Algeria                   89416.0        58146.0      2539.0     28731.0
Andorra                    7127.0         6367.0        78.0       682.0
Angola                    15729.0         8470.0       355.0      6904.0
...                           ...            ...         ...         ...
Vietnam                    1377.0         1224.0        35.0       118.0
West Bank and Gaza       101109.0        75257.0       867.0     24985.0
Yemen                      2078.0         1382.0       606.0        90.0
Zambia                    17963.0        17268.0       364.0       331.0
Zimbabwe                  10912.0         9062.0       303.0      1547.0

[191 rows x 4 columns]

### Import Libraries

In [2]:
import pandas as pd
import numpy as np

zzzz## Load Datasets